tree_test.go

```diff
@@ -1649,3 +1654,127 @@ func TestLeafNodeInsert(t *testing.T) {
 		t.Fatalf("hash should not be nil")
 	}
 }
+
+type randTest []randTestStep
+
+type randTestStep struct {
+	op    int
+	key   []byte // for opInsert, opDelete, opGet
+	value []byte // for opInsert
+	err   error  // for debugging
+}
+
+const (
+	opInsert = iota
+	opDelete
+	opGet
+	opHash
+	opCommit
+	opProve
+	opMax
+)
+
+func (randTest) Generate(r *mRand.Rand, size int) reflect.Value {
+	var finishedFn = func() bool {
+		size--
+		return size == 0
+	}
+	return reflect.ValueOf(generateSteps(finishedFn, r))
+}
+
+func generateSteps(finished func() bool, r io.Reader) randTest {
+	var allKeys [][]byte
+	var one = make([]byte, 32)
+	genKey := func() []byte {
+		r.Read(one)
+		if len(allKeys) < 2 || one[0]%100 > 90 {
+			// new key
+			key := make([]byte, 32)
+			r.Read(key)
+			allKeys = append(allKeys, key)
+			return key
+		}
+		// use existing key
+		idx := int(one[0]) % len(allKeys)
+		return allKeys[idx]
+	}
+	var steps randTest
+	for !finished() {
+		r.Read(one)
+		step := randTestStep{op: int(one[0]) % opMax}
+		switch step.op {
+		case opInsert:
+			step.key = genKey()
+			step.value = make([]byte, 8)
+			binary.BigEndian.PutUint64(step.value, uint64(len(steps)))
```

Comment:
Curious about the logic of using the len of steps, instead of something like `r.Read(step.value)`? And also why `step.value` has length `8` instead of `32`?

Not even sure that changing to what I said would be better really -- just asking if there was a reason that I might be missing on why length is 8 and value is `len(steps)`.

Reply:
I don't think there are any particular reasons that it must be a length of `8` (aside from the fact that we are using `PutUint64`) and use the `len(steps)` as the value, since what it does is just trying to create some arbitrary values.

But I do think that it's better to just create random values by using `r.Read(step.value)`.